enhancements/substitutes-for.md

```diff
@@ -0,0 +1,386 @@
+---
+title: Substitutes-For-Graph-Replacements
+authors:
+  - "@bparees"
+reviewers:
+  - "@ecordell"
+  - "@kevinrizza"
+  - "@lgalleti"
+approvers:
+  - TBD
+creation-date: 2020-12-17
+last-updated: 2020-12-17
+status: provisional
+---
+
+# Bundle Substitution
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+1. Can we get the operator pipeline using a single opm binary that is always the latest
+version (for both DB building and including in catalogs)?  If not, why not? (it means we 
+must backport this feature to all opm versions the operator pipeline uses today)
+
+## Summary
+
+It is currently very difficult to inject a new version of a bundle into the middle of an upgrade graph.  This
+is because existing nodes in the graph may contain explicit "replaces" or "skips" references to other nodes.
+If a new bundle is added to the middle of the graph, replaces/skips references must be updated to include
+the new node.
+
+There is a need to rebuild old versions of bundles to pick up CVE fixes.  Doing so requires that the
+upgrade graph is preserved such that users who update to the CVE-patched bundle can still continue to
+upgrade to newer bundles, just as if they were still on the un-patched bundle.
+
+For example, suppose the original upgrade graph, as specified via "replaces" references, was:
+
+1.0.0->1.0.1
+
+If the patched version is 1.0.1-patched (1.0.1-patched is the 1.0.0 bundle, rebuilt with CVE fixes.  It is not 
+the v1.0.1 version of the operator.  Due to semver semantics this is ordered before 1.0.1, which is what we want), 
+then ideally we want the new graph to be:
+
+[1.0.0 or 1.0.1-patched]->1.0.1
+
+However achieving such a graph would require updating the 1.0.1 bundle to add a "replaces 1.0.1-patched" reference.
+In turn, 1.0.1 can't be updated without changing its version, which means we end up needing something like:
+
+[1.0.0 or 1.0.1-patched or 1.0.1]->1.0.2
+
+(Where the only difference between 1.0.1 and 1.0.2 is the replaces references).
+
+
+To avoid having to update and publish new versions of subsequent bundles in the upgrade graph, we need a way to
+effectively stitch 1.0.1-patched into the graph without changing 1.0.1 (or any other existing bundles that need
+to be able to upgrade from 1.0.1-patched).
+
+This enhancement proposes we do so by allowing a bundle to specify which other bundles it "substitutes for".  This
+will effectively act as a search/replace in the existing bundles for any references to the bundles this bundle 
+substitutes for.  
+
+Continuing the example above, if 1.0.1-patch is annotated as substituting for 1.0.0, then the effective new upgrade 
+graph would be:
+
+1.0.0->1.0.1-patched->1.0.1
```

Comment:
This may be addressed later, but I just want to check. should this be:

1.0.0->1.0.1-patched->1.0.1
1.0.0->1.0.1

Reply:
if you mean "in addition to going from 1.0.0 to 1.0.1-patched to 1.0.1" users can also go directly from 1.0.0 to 1.0.1 still, then yes.  From a CSV perspective, we started with:

1.0.1 replaces 1.0.0

and now we have:
1.0.1 replaces 1.0.1-patched, skips 1.0.0.
1.0.1-patched skips 1.0.0

does that make sense?

i'll add 1.0.0->1.0.1 to this.

Reply:
Yep, exactly - thanks for clarifying